jaclang_jaseci/plugins/common.py

```diff
@@ -0,0 +1,1174 @@
+
+    async def destroy_with_session(self) -> None:
+        """Destroy Architype with session."""
+        async with await ArchCollection.get_session() as session:
+            async with session.start_transaction():
+                try:
+                    await self.destroy(session)
+                    await session.commit_transaction()
+                except Exception:
+                    await session.abort_transaction()
+                    logger.exception(f"Error destroying {self._jac_type_.name}!")
+                    raise
+
+    async def is_allowed(self, to: DA, jctx: Optional["JacContext"] = None) -> bool:
+        """Access validation."""
+        if not jctx:
+            jctx = JCONTEXT.get()
+        if (
+            not jctx
+            or not (from_jd := self._jac_doc_).connected
+            or not (to_jd := to._jac_doc_).connected
+            or not (from_root := from_jd.root)
+            or not (to_root := to_jd.root)
+        ):
+            return False
+
+        if (isinstance(self, Root) and from_jd.id == to_root) or from_root == to_root:
+            to_jd.current_access_level = 1
+            return True
+
+        if (to_access := to_jd.access).all:
+            to_jd.current_access_level = to_access.all - 1
+            return True
+
+        for i in range(1, -1, -1):
+            if from_jd.id in to_access.nodes[i] or from_root in to_access.roots[i]:
+                to_jd.current_access_level = i
+                return True
+
+        if isinstance(
+            to_root_access := await jctx.check_root_access(to_root),
+            DocAccess,
+        ) and (cur_root_id := jctx.get_root_id()):
+            for i in range(1, -1, -1):
+                if cur_root_id in to_root_access.roots[i]:
+                    to_jd.current_access_level = i
+                    return True
+
+        to_jd.current_access_level = None
+        return False
+
+    def __eq__(self, other: object) -> bool:
+        """Override equal implementation."""
+        if isinstance(other, DocArchitype):
+            return self._jac_doc_ == other._jac_doc_
+        elif isinstance(other, DocAnchor):
+            return self._jac_doc_ == other
+
+        return False
+
+    def __hash__(self) -> int:
+        """Override hash implementation."""
+        return self._jac_doc_.__hash__()
+
+
+class NodeArchitype(_NodeArchitype, DocArchitype["NodeArchitype"]):
+    """Overriden NodeArchitype."""
+
+    _jac_type_ = JType.node
+
+    def __init__(self) -> None:
+        """Create node architype."""
+        self._jac_: NodeAnchor = NodeAnchor(obj=self)
+
+    async def connect(self) -> "NodeArchitype":
+        """Return self."""
+        return self
+
+    class Collection(ArchCollection["NodeArchitype"]):
+        """Default NodeArchitype Collection."""
+
+        __collection__ = "node"
+        __indexes__ = [
+            {"fields": [("_id", ASCENDING), ("name", ASCENDING), ("root", ASCENDING)]}
+        ]
```

Comment:
We should document the key-value structure for each architype (node, edge, user) somewhere. Maybe not in the code here but in a documentation markdown somewhere, as part of this PR.

Reply:
will add it